src/library/scanner/libraryscanner.cpp

```diff
@@ -476,6 +510,58 @@ void LibraryScanner::addNewTrack(TrackPointer pTrack) {
         emit(trackAdded(pTrack));
         emit(progressLoading(pTrack->getLocation()));
     } else {
-        qWarning() << "Track ("+pTrack->getLocation()+") could not be added";
+        qWarning()
+                << "Track (" + pTrack->getLocation() + ") could not be added";
+    }
+}
+
+bool LibraryScanner::changeScannerState(ScannerState newState) {
+    switch (newState) {
+    case IDLE:
+        // we are leaving STARTING  or CANCELING state
+        // m_state is already IDLE if a scan was canceled
+        m_state = IDLE;
+        m_stateMutex.unlock();
+        return true;
+    case STARTING:
+        // we need to lock the mutex during the STARTING state
+        // to prevent loosing cancel commands or start the scanner
+        // twice
+        if (m_stateMutex.tryLock()) {
+            if (m_state != IDLE) {
+                qDebug() << "LibraryScanner: Scan already in progress.";
+                m_stateMutex.unlock();
+                return false;
+            }
+            m_state = STARTING;
+            return true;
+        } else {
+            qDebug() << "LibraryScanner: mutex locked, state =" << m_state;
+            return false;
+        }
+    case SCANNING:
+        DEBUG_ASSERT(m_state == STARTING);
+        // Transition protected by the mutex is over now
+        // Allow canceling
+        m_state = SCANNING;
+        m_stateMutex.unlock();
+        return true;
+    case CANCELING:
+        // canceling is always possible, but wait
+        // until there is no scan starting.
+        m_stateMutex.lock();
```

Comment:
The unconventional usage of the mutex was (and still is) my concern. I don't 
remember that I ever used such a pattern when implementing worker threads or 
a thread pool. On the other hand I couldn't find any transition that locks 
or unlocks the mutex twice.

There is a small chance that the mutex does not get unlocked if 
slotStartScan() is left through an exception, because the the mutex is 
unlocked at the end of this function when the state is finally changed to 
SCANNING. @daschuer Would it be possible to change the state earlier to 
SCANNING or IDLE depending on dirs.isEmpty() at the beginning of this function?

On 08/17/2015 07:14 PM, Owen Williams wrote:
>
> In src/library/scanner/libraryscanner.cpp 
> <https://github.com/mixxxdj/mixxx/pull/655#discussion_r37211353>:
>
> > +            return true;
> > +        } else {
> > +            qDebug() << "LibraryScanner: mutex locked, state =" << m_state;
> > +            return false;
> > +        }
> > +    case SCANNING:
> > +        DEBUG_ASSERT(m_state == STARTING);
> > +        // Transition protected by the mutex is over now
> > +        // Allow canceling
> > +        m_state = SCANNING;
> > +        m_stateMutex.unlock();
> > +        return true;
> > +    case CANCELING:
> > +        // canceling is always possible, but wait
> > +        // until there is no scan starting.
> > +        m_stateMutex.lock();
>
> It seems that the lock is being used more like a boolean than a thread 
> lock. The pattern of locking and unlocking in different functions seems to 
> risk deadlocks. What if CANCEL is fired twice? Won't it double lock and hang?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/655/files#r37211353>.
>